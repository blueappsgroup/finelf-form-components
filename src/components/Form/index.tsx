import React, { FC, ReactElement, useMemo, useState } from 'react'
import { Formik, Form } from 'formik'
import styled from 'styled-components'

import { ThemeProvider } from '../../consts/theme'
import { device } from '../../consts/sizes'
import { FormProps, FormValuesType } from '../../types'
import RedirectPage from '../RedirectPage'
import {
  FormContext,
  getFormValuesFromCache,
  resetFormValueCache,
  handleSendDataToApi,
  getFieldsValuesFromUrl,
} from '../../utils'
import { formStatuses } from '../../consts/form'

const StyledForm = styled(Form)`
  display: flex;
  flex-direction: column;
  max-width: ${(props): string => props.theme.formMaxWidth};
  justify-self: center;
  background: ${(props): string => props.theme.formBgColor};
  font-family: ${(props): string => props.theme.fontFamily};
  border-radius: 6px;
  box-shadow: ${(props): string => props.theme.formBoxShadow};
`

const FormWrapper: FC<FormProps> = ({
  children,
  onSubmit,
  customTheme,
  id,
  stepsLength,
  stepsTitles,
  hasRedirect,
  redirectUrl,
  timeToRedirect,
  redirectHeaderText,
  logoImg,
  redirectMainImg,
  redirectBgImg,
  sendDataToApi,
  apiUrl,
  transactionName,
  propertyNamesFromUrl,
  dataWithUserAgent,
}) => {
  const trasationIdValue =
    transactionName &&
    new URLSearchParams(window.location.search).get(transactionName)
  const intialValuesFromUrl = useMemo(
    () =>
      (propertyNamesFromUrl &&
        propertyNamesFromUrl.length > 0 &&
        getFieldsValuesFromUrl(propertyNamesFromUrl)) ||
      {},
    [propertyNamesFromUrl]
  )
  const [redirectUrlPath, setRedirectUrlPath] = useState(redirectUrl)
  const [initialValues, setInitialValues] = useState({
    ...intialValuesFromUrl,
    ...getFormValuesFromCache(id),
    // eslint-disable-next-line @typescript-eslint/camelcase
    trasaction_id: trasationIdValue,
  })
  const [currentStep, setCurrentStep] = useState(0)
  const [fieldsForSkip, setFieldsForSkip] = useState<string[]>([])
  const [errorFromApi, setErrorFromApi] = useState<boolean>(false)
  const shouldRedirect = !errorFromApi && hasRedirect
  const addFieldForSkip = (key: string): void =>
    setFieldsForSkip([...fieldsForSkip, key])

  const handleSubmit = async (
    values: FormValuesType,
    props: {
      resetForm: () => void
      setStatus: (status: string) => void
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ): Promise<any> => {
    if (onSubmit) {
      onSubmit(values, props)
    }

    try {
      if (sendDataToApi && apiUrl) {
        const response = await handleSendDataToApi(
          values,
          apiUrl,
          id,
          fieldsForSkip,
          dataWithUserAgent
        )
        const {
          redirectUrl: urlFromApi,
          status: statusFromApi,
        } = await response.json()

        if (statusFromApi === false) {
          setErrorFromApi(true)
          return
        }

        urlFromApi && setRedirectUrlPath(urlFromApi)
      }

      props.resetForm()
      props.setStatus(formStatuses.submited)
    } catch (e) {
      props.setStatus(formStatuses.error)
    }
  }
  const handleReset = (): void => {
    resetFormValueCache(id)
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    setInitialValues({} as any)
  }

  const prevStep: Function = () => setCurrentStep(currentStep - 1)

  const nextStep: Function = () => setCurrentStep(currentStep + 1)

  return (
    <FormContext.Provider
      value={{
        id,
        apiUrl,
        stepsLength,
        currentStep,
        stepsTitleList: stepsTitles,
        nextStep,
        prevStep,
        fieldsForSkip,
        addFieldForSkip,
        errorFromApi,
      }}
    >
      <ThemeProvider customTheme={{ ...customTheme }}>
        <Formik
          enableReinitialize
          initialValues={initialValues}
          onSubmit={handleSubmit}
          onReset={handleReset}
        >
          {(props): ReactElement =>
            (shouldRedirect && props.status === formStatuses.submited && (
              <RedirectPage
                redirectUrl={redirectUrlPath}
                backgroundImage={redirectBgImg}
                logoImg={logoImg}
                headerText={redirectHeaderText}
                timeToRedirect={timeToRedirect}
                mainImg={redirectMainImg}
              />
            )) || (
              <StyledForm id={id}>
                {children}
              </StyledForm>
            )
          }
        </Formik>
      </ThemeProvider>
    </FormContext.Provider>
  )
}

export default FormWrapper
