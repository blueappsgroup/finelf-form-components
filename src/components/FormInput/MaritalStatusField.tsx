/* eslint-disable @typescript-eslint/camelcase */
import { Field } from 'formik'
import React, { ReactElement } from 'react'
import { SelectFieldProps } from '../../types'

import { BaseSelectField } from './base'

const options = {
  unmarried: 'panna/kawaler',
  married: 'mężatka/żonaty',
  divorced: 'rozwiedziona/rozwiedziony',
  widower: 'wdowa/wdowiec',
  with_parents: 'z rodzicami',
}

const MaritalStatusField: (props: SelectFieldProps) => ReactElement = ({
  id,
  label,
  name = 'marital_status',
  required = false,
  showError,
}) => (
  <Field
    required={required}
    showError={showError}
    id={id}
    name={name}
    component={BaseSelectField}
    label={label}
    type="select"
    options={options}
  />
)

export default MaritalStatusField
