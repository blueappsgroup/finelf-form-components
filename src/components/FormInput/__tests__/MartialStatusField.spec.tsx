import * as React from 'react'
import { render, fireEvent, act } from '@testing-library/react'

import Form from '../../Form'
import MaritalStatusField from '../MaritalStatusField'

describe('<MaritalStatusField />', () => {
  const onSubmit = jest.fn()
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const setupWrapper: (config: any) => any = ({
    formId = 'testForm',
    inputId = 'maritalStatusField',
    inputName = 'name',
    required = false,
    ...rest
  }) => {
    const wrapper = render(
      <Form id={formId} onSubmit={onSubmit}>
        <MaritalStatusField
          id={inputId}
          name={inputName}
          required={required}
          {...rest}
        />
      </Form>
    )
    const { container } = wrapper

    const input = container.querySelector(`[name="${inputName}"]`)

    return {
      input,
      ...wrapper,
    }
  }

  it('matches snapshot', () => {
    const { input } = setupWrapper({})

    expect(input).toBeTruthy()
    expect(input).toMatchSnapshot()
  })

  it('matches snapshot with label', () => {
    const wrapper = setupWrapper({ label: 'Test' })
    const { container } = wrapper

    expect(wrapper.getByText('Test')).toBeTruthy()
    expect(container).toMatchSnapshot()
  })

  it('changes input field', async () => {
    const { input } = setupWrapper({})

    await act(async () => {
      fireEvent.change(input, { target: { value: 'married' } })
    })

    expect(input.value).toBe('married')
  })

  it('input field onBlur save value to sessionStorage', async () => {
    const { input } = setupWrapper({})

    await act(async () => {
      fireEvent.change(input, { target: { value: 'married' } })
    })

    await act(async () => {
      fireEvent.blur(input, { target: { value: 'married' } })
    })

    expect(
      JSON.parse(global.window.sessionStorage.getItem('form-testForm')).name
    ).toBe('married')
    expect(input.value).toBe('married')
  })
})
