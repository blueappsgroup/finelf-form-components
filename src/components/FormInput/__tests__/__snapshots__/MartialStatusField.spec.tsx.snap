// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MaritalStatusField /> matches snapshot 1`] = `
<select
  class="sc-dIUggk hhFmtN"
  id="maritalStatusField"
  name="name"
  type="text"
>
  <option
    class="sc-dmlrTW hequYp"
    disabled=""
    value="select"
  >
    wybierz 
  </option>
  <option
    value="unmarried"
  >
    panna/kawaler
  </option>
  <option
    value="married"
  >
    mężatka/żonaty
  </option>
  <option
    value="divorced"
  >
    rozwiedziona/rozwiedziony
  </option>
  <option
    value="widower"
  >
    wdowa/wdowiec
  </option>
  <option
    value="with_parents"
  >
    z rodzicami
  </option>
</select>
`;

exports[`<MaritalStatusField /> matches snapshot with error 1`] = `
<div>
  <form
    action="#"
    class="sc-fubCfw clDFWv"
    id="testForm"
  >
    <div
      class="sc-iqHYGH jWqqni"
    >
      <label
        for="name"
      >
        Test*
      </label>
      <div
        class="sc-idOhPF buAmNl"
      >
        <select
          class="sc-dIUggk jigzeI"
          id="maritalStatusField"
          label="Test"
          name="name"
          required=""
          type="text"
        >
          <option
            class="sc-dmlrTW hequYp"
            disabled=""
            value="select"
          >
            wybierz *
          </option>
          <option
            value="unmarried"
          >
            panna/kawaler
          </option>
          <option
            value="married"
          >
            mężatka/żonaty
          </option>
          <option
            value="divorced"
          >
            rozwiedziona/rozwiedziony
          </option>
          <option
            value="widower"
          >
            wdowa/wdowiec
          </option>
          <option
            value="with_parents"
          >
            z rodzicami
          </option>
        </select>
        <span
          class="sc-hHftDr jmpDvw"
        />
      </div>
      <span
        class="sc-hBEYos kIOtFs"
      >
        To pole jest wymagane
      </span>
    </div>
    <button
      class="sc-iBPRYJ hcYpVh"
      disabled=""
      type="submit"
    >
      Wyślij
    </button>
  </form>
</div>
`;

exports[`<MaritalStatusField /> matches snapshot with label 1`] = `
<div>
  <form
    action="#"
    class="sc-fubCfw clDFWv"
    id="testForm"
  >
    <div
      class="sc-iqHYGH jWqqni"
    >
      <label
        for="name"
      >
        Test
      </label>
      <div
        class="sc-idOhPF buAmNl"
      >
        <select
          class="sc-dIUggk hhFmtN"
          id="maritalStatusField"
          label="Test"
          name="name"
          type="text"
        >
          <option
            class="sc-dmlrTW hequYp"
            disabled=""
            value="select"
          >
            wybierz 
          </option>
          <option
            value="unmarried"
          >
            panna/kawaler
          </option>
          <option
            value="married"
          >
            mężatka/żonaty
          </option>
          <option
            value="divorced"
          >
            rozwiedziona/rozwiedziony
          </option>
          <option
            value="widower"
          >
            wdowa/wdowiec
          </option>
          <option
            value="with_parents"
          >
            z rodzicami
          </option>
        </select>
        <span
          class="sc-hHftDr jmpDvw"
        />
      </div>
    </div>
    <button
      class="sc-iBPRYJ hcYpVh"
      type="submit"
    >
      Wyślij
    </button>
  </form>
</div>
`;

exports[`<MaritalStatusField /> select field with custom validation 1`] = `
<div>
  <form
    action="#"
    class="sc-fubCfw clDFWv"
    id="testForm"
  >
    <div
      class="sc-iqHYGH jWqqni"
    >
      <div
        class="sc-idOhPF buAmNl"
      >
        <select
          class="sc-dIUggk jigzeI"
          id="maritalStatusField"
          name="name"
          type="text"
        >
          <option
            class="sc-dmlrTW hequYp"
            disabled=""
            value="select"
          >
            wybierz 
          </option>
          <option
            value="unmarried"
          >
            panna/kawaler
          </option>
          <option
            value="married"
          >
            mężatka/żonaty
          </option>
          <option
            value="divorced"
          >
            rozwiedziona/rozwiedziony
          </option>
          <option
            value="widower"
          >
            wdowa/wdowiec
          </option>
          <option
            value="with_parents"
          >
            z rodzicami
          </option>
        </select>
        <span
          class="sc-hHftDr jmpDvw"
        />
      </div>
      <span
        class="sc-hBEYos kIOtFs"
      >
        Podane dane są nieprawidłowe
      </span>
    </div>
    <button
      class="sc-iBPRYJ hcYpVh"
      disabled=""
      type="submit"
    >
      Wyślij
    </button>
  </form>
</div>
`;
